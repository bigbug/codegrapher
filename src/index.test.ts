import getDigraph from ".";
const fs = require('fs');

//const input = 'if(b<0) {\nb=-b;\n}\nb=2+b;\nif(a>1&&a<10&&a==2) {b=10;} else {b=a;} b=b * 2;c=a-b;';
//const input = 'int a = a+2;';
//const input = 'a=2+b;';
//const input = 'if(b==2) {b=10;} else {b=2*b;}';

let input = "";
try {
    input = fs.readFileSync("source.c", "utf-8");
} catch (err) {
    console.error(err);
}


test("a=2+b;", () => {
    const res = getDigraph("a=2+b;");
    expect(res).toMatchSnapshot();
});

test("a=2+(b);", () => {
    const res = getDigraph("a=2+(b);");
    expect(res).toMatchSnapshot();
});

test("if(b==2) {b=10;} else {b=2*b;}", () => {
    const res = getDigraph("if(b==2) {b=10;} else {b=2*b;}");
    expect(res).toMatchSnapshot();
});
test("a=!b;", () => {
    const res = getDigraph("a=!b;");
    expect(res).toMatchSnapshot();
});

test("if(b==2) {a=2;} else {if(b==1) {a=1;} else {a=0;}}", () => {
    const res = getDigraph("if(b==2) {a=2;} else {if(b==1) {a=1;} else {a=0;}}");
    expect(res).toMatchSnapshot();
})

test("b=12;if(a==2) {c=b+34;} else {c=5678;}", () => {
    const res = getDigraph("b=12;if(a==2) {c=b+34;} else {c=5678;}");
    //console.log(res);
    expect(res).toMatchSnapshot();
})

test("arbitrary input", () => {
    const res = getDigraph(input);
    //console.log(res);
    try {
        fs.writeFileSync('reverselinktargetresult.dot', res)
        //file written successfully
      } catch (err) {
        console.error(err)
      }
});