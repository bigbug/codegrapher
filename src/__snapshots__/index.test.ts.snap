// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a=!b; 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n3 [label=\\"b\\"]
node_n2 [label=\\"not\\"]
node_n3 -> node_n2
}"
`;

exports[`a=(b*c)-h; 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"b\\"]
node_n3 [label=\\"c\\"]
node_n4 [label=\\"*\\\\n*\\"]
node_n5 [label=\\"h\\"]
node_n6 [label=\\"+\\\\n-\\"]
node_n2 -> node_n4
node_n3 -> node_n4
node_n4 -> node_n6
node_n5 -> node_n6
}"
`;

exports[`a=2+(b); 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"2\\"]
node_n3 [label=\\"b\\"]
node_n4 [label=\\"+\\\\n+\\"]
node_n2 -> node_n4
node_n3 -> node_n4
}"
`;

exports[`a=2+b; 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"2\\"]
node_n3 [label=\\"b\\"]
node_n4 [label=\\"+\\\\n+\\"]
node_n2 -> node_n4
node_n3 -> node_n4
}"
`;

exports[`b=12;if(a==2) {c=b+34;} else {c=5678;} 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"12\\"]
node_n4 [label=\\"34\\"]
node_n5 [label=\\"+\\\\n+\\"]
node_n7 [label=\\"a\\"]
node_n8 [label=\\"2\\"]
node_n6 [label=\\"==\\"]
node_n9 [label=\\"5678\\"]
node_n3 [label=\\"1\\\\ncond\\\\n2\\"]
node_n2 -> node_n5 [label=\\"b\\"]
node_n4 -> node_n5
node_n7 -> node_n6
node_n8 -> node_n6
node_n5 -> node_n3
node_n6 -> node_n3
node_n9 -> node_n3
}"
`;

exports[`function declaraton 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]

subgraph cluster_n2 {label=\\"(int)max\\"
node_n3 [label=\\"num1\\" color=orange]
node_n4 [label=\\"num2\\" color=orange]
node_n6 [label=\\">\\"]
node_n5 [label=\\"1\\\\ncond\\\\n2\\"]
node_n7 [label=\\"return\\" color=orange]
node_n3 -> node_n6 [label=\\"num1\\"]
node_n4 -> node_n6 [label=\\"num2\\"]
node_n3 -> node_n5 [label=\\"num1\\"]
node_n6 -> node_n5
node_n4 -> node_n5 [label=\\"num2\\"]
node_n5 -> node_n7 [label=\\"result\\"]
}}"
`;

exports[`if pipe with scope 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"2\\"]
node_n3 [label=\\"1\\"]
node_n4 [label=\\"+\\\\n+\\"]
node_n6 [label=\\"2\\"]
node_n5 [label=\\"==\\"]
node_n20 [label=\\"pipe\\"]
node_n22 [label=\\"b\\"]
node_n21 [label=\\"pipe\\"]
subgraph cluster_n7 {node_n8 [label=\\"1\\"]
node_n9 [label=\\"3\\"]
node_n11 [label=\\"3\\"]
node_n10 [label=\\"==\\"]
node_n16 [label=\\"pipe\\"]
node_n17 [label=\\"pipe\\"]
node_n18 [label=\\"+\\\\n+\\"]
node_n19 [label=\\"act\\" color=green]
subgraph cluster_n12 {node_n13 [label=\\"2\\"]
node_n14 [label=\\"2\\"]
node_n15 [label=\\"act\\" color=green]
node_n10 -> node_n15
}node_n9 -> node_n10 [label=\\"a\\"]
node_n11 -> node_n10
node_n15 -> node_n16
node_n8 -> node_n16 [label=\\"c\\"]
node_n13 -> node_n16
node_n15 -> node_n17
node_n9 -> node_n17 [label=\\"a\\"]
node_n14 -> node_n17
node_n17 -> node_n18 [label=\\"a\\"]
node_n16 -> node_n18 [label=\\"c\\"]
node_n5 -> node_n19
}node_n2 -> node_n4 [label=\\"a\\"]
node_n3 -> node_n4
node_n4 -> node_n5 [label=\\"a\\"]
node_n6 -> node_n5
node_n19 -> node_n20
node_n4 -> node_n20 [label=\\"a\\"]
node_n17 -> node_n20
node_n19 -> node_n21
node_n22 -> node_n21
node_n18 -> node_n21
}"
`;

exports[`if(b==2) {a=2;} else {if(b==1) {a=1;} else {a=0;}} 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n3 [label=\\"2\\"]
node_n5 [label=\\"b\\"]
node_n6 [label=\\"2\\"]
node_n4 [label=\\"==\\"]
node_n7 [label=\\"a\\"]
node_n2 [label=\\"1\\\\ncond\\\\n2\\"]
node_n5 -> node_n4
node_n6 -> node_n4
node_n3 -> node_n2
node_n4 -> node_n2
node_n7 -> node_n2
}"
`;

exports[`if(b==2) {b=10;} else {b=2*b;} 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n3 [label=\\"10\\"]
node_n5 [label=\\"b\\"]
node_n6 [label=\\"2\\"]
node_n4 [label=\\"==\\"]
node_n7 [label=\\"2\\"]
node_n8 [label=\\"*\\\\n*\\"]
node_n2 [label=\\"1\\\\ncond\\\\n2\\"]
node_n5 -> node_n4
node_n6 -> node_n4
node_n7 -> node_n8
node_n5 -> node_n8
node_n3 -> node_n2
node_n4 -> node_n2
node_n8 -> node_n2
}"
`;

exports[`post decrement 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"i\\"]
node_n4 [label=\\"1\\"]
node_n3 [label=\\"+\\\\n-\\"]
node_n2 -> node_n3
node_n4 -> node_n3
}"
`;

exports[`post increment 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"i\\"]
node_n4 [label=\\"1\\"]
node_n3 [label=\\"+\\\\n+\\"]
node_n2 -> node_n3
node_n4 -> node_n3
}"
`;

exports[`pre decrement 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"i\\"]
node_n4 [label=\\"1\\"]
node_n3 [label=\\"+\\\\n-\\"]
node_n2 -> node_n3
node_n4 -> node_n3
}"
`;

exports[`pre increment 1`] = `
"digraph {
node [shape=box style=filled]
graph [rankdir=\\"LR\\"]
node_n2 [label=\\"i\\"]
node_n4 [label=\\"1\\"]
node_n3 [label=\\"+\\\\n+\\"]
node_n2 -> node_n3
node_n4 -> node_n3
}"
`;
